### PR TITLE
mon/MDSMonitor: wait for readable OSDMap before sanitizing

### PR DESCRIPTION
**Pull request by batrick**
Fixes: http://tracker.ceph.com/issues/21945

Signed-off-by: Patrick Donnelly <pdonnell@redhat.com>

### CODE REVIEW

File: src/mon/MDSMonitor.cc [hunk 1]
```diff
@@ -137,6 +135,11 @@ void MDSMonitor::init()
 
 void MDSMonitor::create_pending()
 {
+  if (mon->osdmon()->is_readable()) {
+    auto &osdmap = mon->osdmon()->osdmap;
+    fsmap.sanitize([&osdmap](int64_t pool){return osdmap.have_pg_pool(pool);});
```
Comment (liewegas) [inline src/mon/MDSMonitor.cc]: s/fsmap/pending_fsmap/

**Comment by liewegas** [comment]
retest this please

File: src/mon/MDSMonitor.cc [hunk 2]
```diff
@@ -140,6 +138,11 @@ void MDSMonitor::create_pending()
   pending_fsmap = fsmap;
   pending_fsmap.epoch++;
 
+  if (mon->osdmon()->is_readable()) {
```
Comment (tchaikov) [inline src/mon/MDSMonitor.cc]: if you have concern about the ordering in which osdmap is updated and mdsmap is updated, as`PAXOS_MDSMAP < PAXOS_OSDMAP`. probably you can do this in `post_paxos_update()` where all paxos services have been refreshed.

Comment (liewegas) [reply src/mon/MDSMonitor.cc]: that ensures a valid osdmap, but we still shouldn't touch the fsmap.  the change should be applied to pending_fsmap, which means it goes in create_pending or encode_pending

Comment (tchaikov) [reply src/mon/MDSMonitor.cc]: ahh, agreed.

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-10-30_14:46:06-rados-wip-sage-testing-2017-10-30-0722-distro-basic-smithi/

**Pull request merged**